### PR TITLE
Resolve __DIR__ / __FILE__ when const/variable is used for include

### LINKED ISSUE #7584: UnresolvableInclude when include uses __DIR__ assigned to const or variable
__DIR__ and __FILE__ aren't resolved when const/variable is used by include.

https://psalm.dev/r/669175409c

### PR DESCRIPTION
**Pull request by tomasz-kusy**
Fixes #7584

### CODE REVIEW

File: src/Psalm/Internal/Analyzer/Statements/Expression/MagicConstAnalyzer.php [hunk 1]
```diff
@@ -84,10 +84,16 @@ public static function analyze(
...
                 $statements_analyzer->node_data->setType($stmt, new Union([new TCallableString]));
             }
-        } elseif ($stmt instanceof PhpParser\Node\Scalar\MagicConst\File
-            || $stmt instanceof PhpParser\Node\Scalar\MagicConst\Dir
-        ) {
-            $statements_analyzer->node_data->setType($stmt, new Union([new TNonEmptyString()]));
+        } elseif ($stmt instanceof PhpParser\Node\Scalar\MagicConst\Dir) {
+            $statements_analyzer->node_data->setType(
+                $stmt,
+                Type::getString(\dirname($statements_analyzer->getSource()->getFilePath()))
```
Comment (weirdan) [inline src/Psalm/Internal/Analyzer/Statements/Expression/MagicConstAnalyzer.php:90]: Please import `dirname()` with `use function dirname;` and drop the backslash.

**Review by weirdan — CHANGES_REQUESTED** [review]

**Comment by orklah** [comment]
Thanks! Congrats for your first PR!

**Review by weirdan — APPROVED** [review]

**Comment by weirdan** [comment]
Thanks!

**Comment by VincentLanglet** [comment]
Now `__DIR__` is resolved, I ends with multiple errors similar to the issue
https://github.com/vimeo/psalm/issues/3886 with the code
```
if (file_exists($file = __DIR__.'/custom_bootstrap.php')) {
    require_once $file;
}
```
Wouldn't it be possible for psalm to not report an error in this situation ?

Since the file with the code is autogenerated, and sometimes there is a custom_bootstrap, sometimes there isn't, I cannot add a psalm-suppress comment (because it will sometimes be reported as unused). 
Any recommendation in this situation @orklah ?

**Comment by orklah** [comment]
Probably related to https://github.com/vimeo/psalm/issues/3241 and https://github.com/vimeo/psalm/issues/4788

If the expression is too complex, Psalm can't see you're requiring the same thing.

If you have control over the code generation, consider introducing a local variable before the file_exists, this may help

**Pull request merged**